### ISSUE TITLE
parallel tests each creating an async server stomp

### ISSUE THREAD

**Issue by evbo** [lipanski/mockito#200]
It seems if my tests run in parallel and both define a `ServerGuard` the mocks stomp one another.

The only solution I've found is running them serially.

How can multiple tests create a server like?:
```rust
        let mut server = mockito::Server::new_async().await;
        server.mock(..)
```

without affecting other tests?

**Comment by lipanski** [comment]
@evbo can you provide some code to reproduce this? this repo runs tests in parallel and creates lots of server guards without any knows issues so far.

**Comment by lipanski** [comment]
Also note that there's a [limit](https://github.com/lipanski/mockito/blob/master/src/server_pool.rs#L9-L10) to the amount of servers you can run in parallel (20 for mac, 50 otherwise) and if you want more you should just use `Server::new_with_opts`, which bypasses the server pool.

**Comment by evbo** [comment]
Sorry I couldn't get to this in time. I will try `new_with_opts` as my tests are definitely stomping and I only have 2 servers! May be a while when I get around to it... Like I said everything works great running tests one at a time at least.

**Comment by evbo** [comment]
@lipanski can you please reopen?

I think I found the issue. It appears ServerGuard does not support having multiple paths that are subsets, e.g.:

```rust
server
  .mock("POST", "/uri/hello")
  .with_status(200)
  .create();

// if I don't comment this I get not implemented errors when accesssing customizations
let id = "123";
server
  .mock("GET", &*format!("/uri/hello/{id}/stuff"))
  .with_status(200)
  .create();
```

Why would this be the case?

**Comment by evbo** [comment]
`new_with_opts_async` makes no difference. This issue continually occurs on projects where I am testing using `tokio-test`

I suggest leaving it open. It's real. Sorry I can't post my code, but here's a dummy version:

```rust
#[cfg(test)]
mod operations_tests {

    use crate::schema::ModelStatus::{Available, Pending};
    use mockito::Matcher;
    use mockito::ServerOpts;
    use serde::Serialize;
    use std::{env, str::FromStr};
    use tokio::sync::AcquireError;
    use tokio_test as _;
    use AsrError::AudioMissing;
    use AsrError::CannotOverwriteResource;

    use super::super::*;

    fn overwrite_error() -> AsrError {
        AsrError::CannotOverwriteResource {
            id: Uuid::nil().to_string(),
        }
    }

    pub fn json_bytes<T>(structure: T) -> Vec<u8>
    where
        T: Serialize,
    {
        let mut bytes: Vec<u8> = Vec::new();
        serde_json::to_writer(&mut bytes, &structure).unwrap();
        bytes
    }

    #[test]
    fn abnf_vocab_test() {
        let phrases: Vec<String> = [
            "Antonio",
            "Gómez",
            "vendedor",
            "Isabel",
            "Flores",
            "maestra",
            "José",
            "Morales",
            "camionero",
            "Lucía",
            "Castillo",
            "abogada",
            "David",
            "González",
            "banquer",
            "María",
            "Pineda",
            "doctora",
        ]
        .into_iter()
        .map(|s| s.to_owned())
        .collect::<Vec<String>>();

        let abnf = abnf_vocab(&Language::EsLa, &phrases);

        assert!(abnf.starts_with("#ABNF 1.0 ISO-8859-1;"));
    }

    #[tokio::test]
    async fn create_new_model_test() {
        let (mut model, _, client, _server) = mock(&existing_key()).await;

        assert!(matches!(
            create_new_model(&client, "", &mut model).await.unwrap_err(),
            AsrError::CannotOverwriteResource { .. }
        ));

        model.vocab = "new".to_owned();
        assert!(create_new_model(&client, "", &mut model)
            .await
            .unwrap()
            .is_some());
    }

    #[tokio::test]
    async fn add_vocab_to_model_test() {
        let key = existing_key();

        let (_, mut vocab, client, _server) = mock(&key).await;

        assert!(matches!(
            add_vocab_to_model(&client, "token", &key, &vocab)
                .await
                .unwrap_err(),
            CannotOverwriteResource { .. }
        ));

        vocab.name = format!("not{}", vocab.name);
        assert!(add_vocab_to_model(&client, "token", &key, &vocab)
            .await
            .is_ok());
    }


    #[tokio::test]
    async fn train1_test() {
        let (client, _server) = mock().await;
        assert!(..);
    }

    #[tokio::test]
    async fn train2_test() {
        let (client, _server) = mock().await;
        assert!(..);
    }

    async fn mock() -> (reqwest::Client, mockito::Server) {
        
        let mut server = mockito::Server::new_with_opts_async(ServerOpts::default()).await;
        let client = http_client();

        server
            .mock("GET", "/base/logic")
            .with_status(200)
            .create();

        server
            .mock("POST", "/base/logic")
            .with_status(200)
            .create();

        server
            .mock("GET", &*format!("/base/logic/{id}/saved"))
            .with_status(200)
            .create();

        server
            .mock(
                "POST",
                Matcher::Regex(format!("^/base/logic/{id}/saves/[^/]+$")),
            )
            .with_status(200)
            .with_body("{}")
            .create();

        server
            .mock("POST", &*format!("/base/logic/{id}/save"))
            .with_status(200)
            .with_body("{}")
            .create();

        server
            .mock("GET", &*format!("/base/logic/{id}"))
            .with_status(200)
            .create();

        (client, server)
    }
}

```

**Comment by evbo** [comment]
`new_with_opts_async` makes no difference. This issue continually occurs on projects where I am testing using `tokio-test`

I suggest leaving it open. It's real. Sorry I can't post my code, but here's a dummy version:

```rust
#[cfg(test)]
mod tests {
    use mockito::Matcher;
    use mockito::ServerOpts;
    use serde::Serialize;
    use std::{env, str::FromStr};
    use tokio::sync::AcquireError;
    use tokio_test as _;

    use super::super::*;

    #[tokio::test]
    async fn train1_test() {
        let (client, _server) = mock().await;
        assert!(..);
    }

    #[tokio::test]
    async fn train2_test() {
        let (client, _server) = mock().await;
        assert!(..);
    }

    async fn mock() -> (reqwest::Client, mockito::Server) {
        
        let mut server = mockito::Server::new_with_opts_async(ServerOpts::default()).await;
        let client = http_client();

        server
            .mock("GET", "/base/logic")
            .with_status(200)
            .create();

        server
            .mock("POST", "/base/logic")
            .with_status(200)
            .create();

        server
            .mock("GET", &*format!("/base/logic/{id}/saved"))
            .with_status(200)
            .create();

        server
            .mock(
                "POST",
                Matcher::Regex(format!("^/base/logic/{id}/saves/[^/]+$")),
            )
            .with_status(200)
            .with_body("{}")
            .create();

        server
            .mock("POST", &*format!("/base/logic/{id}/save"))
            .with_status(200)
            .with_body("{}")
            .create();

        server
            .mock("GET", &*format!("/base/logic/{id}"))
            .with_status(200)
            .create();

        (client, server)
    }
}

```

**Comment by lipanski** [comment]
@evbo you should use the async methods if you’re going async: `create_async().await` instead of `create`. see the async section of the docs for reference. let me know if that fixes your issue.

**Comment by evbo** [comment]
Thank you @lipanski but it makes no difference. Not only that, if I `println!` my code judiciously I can see the `_server` is being shared between tests, such that it has values from prior tests in the mocked paths.

So you definitely have a bug stemming from some kind of shared mutability between tests. I highly recommend reopening this issue.

**Comment by lipanski** [comment]
@evbo can you share some code to reproduce this? I can't really leave an issue open if I can't reproduce the issue. might be on your side, might be our side - but we can only know once you share some more information. maybe simplify your logic to the point where it's just mockito code and some basic abstraction of what you're doing, including your `println!` stuff. could also be a good exercise to track down any potential issues on your end.

**Comment by weinberggithub** [comment]
same problem, with error Mock not found

**Comment by ChristianPavilonis** [comment]
I'm having the same issue.
I did printlns on the server url and they do have different hosts, but it seems like when one test ends (the server goes out of scope) it cleans up all the mocks so the next one will fail.


```rust

async fn setup() -> mockito::ServerGuard {
    let server = mockito::Server::new_async().await;

    set_host(server.url());

    server
}

#[tokio::test]
async fn test_a() {
    let server = setup().await;
    let mock = server
        .mock("GET", "/api/a")
        .with_status(200)
        .create_async()
        .await;

    call_api_a().await;

    mock.async_assert().await;
}

#[tokio::test]
async fn test_b() {
    let server = setup().await;
    let mock = server
        .mock("GET", "/api/b")
        .with_status(200)
        .create_async()
        .await;

    call_api_b().await;

    mock.assert_async().await;
}
```
